pkg/agent/route/route_linux.go

```diff
@@ -773,3 +856,317 @@ func (c *Client) DeleteSNATRule(mark uint32) error {
+			c.nodeRoutes.Store(routeKey, []*netlink.Route{newRoute})
+		} else {
+			klog.V(4).Infof("Current route can route ClusterIP %v to Antrea gateway", svcIP)
+		}
+	} else {
+		// The route doesn't exist, create one.
+		var mask int
+		var gw *net.IP
+		if isIPv6 {
+			mask = ipv6AddrLength
+			gw = &config.ServiceGWHairpinIPv6
+		} else {
+			mask = ipv4AddrLength
+			gw = &config.ServiceGWHairpinIPv4
+		}
+
+		linkIndex := c.nodeConfig.GatewayConfig.LinkIndex
+		route, err := generateOnlinkRoute(&svcIP, mask, gw, linkIndex, defaultRouteTable)
+		if err != nil {
+			return fmt.Errorf("failed to generate new route %s", svcIP.String())
+		}
+		if err = netlink.RouteReplace(route); err != nil {
+			return fmt.Errorf("failed to install new ClusterIP route: %w", err)
+		}
+
+		c.nodeRoutes.Store(routeKey, []*netlink.Route{route})
+	}
+
+	return nil
+}
+
+// addLoadBalancerIngressIPRoute is used to add routing entry which is used to route LoadBalancer ingress IP to Antrea
+// gateway on host.
+func (c *Client) addLoadBalancerIngressIPRoute(svcIP *net.IP, isIPv6 bool) error {
+	linkIndex := c.nodeConfig.GatewayConfig.LinkIndex
+	var gw *net.IP
+	var mask int
+	if !isIPv6 {
+		gw = &config.ServiceGWHairpinIPv4
+		mask = ipv4AddrLength
+	} else {
+		gw = &config.ServiceGWHairpinIPv6
+		mask = ipv6AddrLength
+	}
+
+	route, err := generateOnlinkRoute(svcIP, mask, gw, linkIndex, defaultRouteTable)
+	if err != nil {
+		return fmt.Errorf("failed to generate routing entry for LoadBalancer ingress IP %s: %w", svcIP.String(), err)
+	}
+	if err = netlink.RouteReplace(route); err != nil {
+		return fmt.Errorf("failed to install routing entry for LoadBalancer ingress IP %s: %w", svcIP.String(), err)
+	}
+	klog.V(4).Infof("Adding LoadBalancer ingress IP route %v", route)
+	c.nodeRoutes.Store(svcIP.String(), []*netlink.Route{route})
+
+	return nil
+}
+
+// deleteLoadBalancerIngressIPRoute is used to delete routing entry which is used to route LoadBalancer ingress IP to Antrea
+// gateway on host.
+func (c *Client) deleteLoadBalancerIngressIPRoute(svcIP *net.IP, isIPv6 bool) error {
+	linkIndex := c.nodeConfig.GatewayConfig.LinkIndex
+	var gw *net.IP
+	var mask int
+	if !isIPv6 {
+		gw = &config.ServiceGWHairpinIPv4
+		mask = 32
+	} else {
+		gw = &config.ServiceGWHairpinIPv6
+		mask = 128
+	}
+
+	route, err := generateOnlinkRoute(svcIP, mask, gw, linkIndex, defaultRouteTable)
+	if err != nil {
+		return fmt.Errorf("failed to generate routing entry for LoadBalancer ingress IP %s: %w", svcIP.String(), err)
+	}
+	if err = netlink.RouteDel(route); err != nil {
+		return fmt.Errorf("failed to delete routing entry for LoadBalancer ingress IP %s: %w", svcIP.String(), err)
+	}
+	klog.V(4).Infof("Deleting LoadBalancer ingress IP route %v", route)
+	c.nodeRoutes.Delete(svcIP.String())
+
+	return nil
+}
+
+// AddLoadBalancer is used to add Linux TC filters and routing entries when a LoadBalancer Service is added.
```

Comment:
Revise the comments and remove TC. The same for other funcs.

Reply:
OK